Default Eigen version set in CMakeLists is not compatible with C++20

**Describe the bug**
Building RDKit from scratch where Eigen is cloned during the build by cmake raises
```
    /Users/runner/work/rdkit-pypi/rdkit-pypi/build/temp.macosx-10.9-x86_64-cpython-39/rdkit/External/Eigen/eigen/Eigen/src/Core/util/Meta.h:320:20: error: no template named 'result_of' in namespace 'std'; did you mean 'boost::result_of'?
      typedef typename std::result_of<T>::type type1;
                       ^~~~~
    /Users/runner/.conan2/p/b/boost515d5c6f3bf28/p/include/boost/utility/result_of.hpp:72:29: note: 'boost::result_of' declared here
    template<typename F> struct result_of;
```

**To Reproduce**
Perform your regular `cmake --build` without doing a conda or apt-get or yum or brew install eigen first.

**Expected behavior**
Build proceeds past the building of Eigen

**Configuration (please complete the following information):**
 - RDKit version: 2025_3_02
 - OS: macOS 13.7.5 (github runner)
 - Python version (if relevant): n/a
 - Are you using conda? no
 - If you are using conda, which channel did you install the rdkit from? n/a
 - If you are not using conda: how did you install the RDKit? n/a

**Additional context**
Fixed in version 3.4.0, by this commit
https://gitlab.com/libeigen/eigen/-/commit/a31effc3bca2f0924752caeebfd6f61f7edf9a43

--- a/External/Eigen/CMakeLists.txt
+++ b/External/Eigen/CMakeLists.txt
@@ -1,5 +1,5 @@
 include(ExternalProject)
-set(EIGEN3_VERSION "3.3.9")
+set(EIGEN3_VERSION "3.4.0")
 set(EIGEN3_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/eigen)
 ExternalProject_Add(Eigen
   GIT_REPOSITORY https://gitlab.com/libeigen/eigen.git

